src/Init/Data/Char/UnicodeSkipList.lean

```diff
@@ -0,0 +1,101 @@
+/-
+Copyright (c) 2024 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jean-Baptiste Tristan
+-/
+prelude
+import Init.Data.Range
+
+/-!
+The unicode skip list data structure is used to represent the set of characters
+that have some property. The data structure is precomputed using the
+tool in script/unicode.
+
+We use a variant of the skip list data structure that takes advantage
+of the distribution of character properties for the common properties of
+interest. Considering the sequence of codepoints that satisfy a given
+property, they tend to appear is small ranges that are clustered. These
+clusters are separated from each other by large ranges of codepoints that
+do not satisfy the property. We refer to such clusters as runs.
+
+The data structure takes advantage of this by representing the set
+using an array of 32 bit integers (the outer array that allows to skip)
+and an array of 8 bit integers. The integers of the outer array
+contains enough information to let us locate in which run a codepoint
+belong, and we call it `runs`. The bytes of the inner array, which we call
+`offsets` record the length of the ranges. Large ranges of codepoints
+that do not satisfy a property appear in the offsets as 0.
+
+A key property of the inner array is that the index of range tells us
+whether the property holds for that range or not. Therefore, to find
+whether a character has a property, we first use the `runs` array
+to locate where to search in the `offsets` array. We can then locate the
+character in the offsets array to determine the parity of its position.
+
+The 32 bit integers in the outer array encode the following
+information. The 11 most significants bits are an index into the inner
+array, pointing to the start of a run. The 21 least significant bits
+point to the codepoint one past the run.
+
+As an example, consider some property that would hold for the set of
+codepoints: {100, 101, 103, 104, 105, 108, 109, 2000, 2001, 2002, 2004}
+
+The offsets would be: [100,2,1,3,2,2,0,3,1,1,0]
+Before packing into a 32 bit integer, the runs would be: [(0,2000),(5,1114112)]
```

Comment:
I think it would be good if the example above already included something about the last 21 bit number always being `1114112` (that is true, right?)